Only call rocblas_initialize for versions < 4 to eliminate unncessary VRAM allocation on some AMD cards

sARY77 opened the pull request:
I have two identical AMD GPUs and noticed the discrepancy in the free memory reported.
I traced it down to the rocblas_initialize call after which the VRAM usage on one of the GPUs jumps up by 498 MiB.

PR that introduced the workaround:
[ROCm Port](https://github.com/ggerganov/llama.cpp/pull/1087)

According to the discussion of the ROCm issue that made the workaround necessary, it has been resolved a long time ago:
[[Bug]: Incorrect results when using GPUs with different architectures](https://github.com/ROCm/rocBLAS/issues/1346)

I tested my change on a model that offloads about 20 GiB on each of the GPUs and did not notice any differences.
I also ran the CI and there and there we no failures reported.

Before:
llama_load_model_from_file: using device ROCm0 (Radeon RX 7900 XTX) - 24026 MiB free
llama_load_model_from_file: using device ROCm1 (Radeon RX 7900 XTX) - 24524 MiB free
After:
llama_load_model_from_file: using device ROCm0 (Radeon RX 7900 XTX) - 24524 MiB free
llama_load_model_from_file: using device ROCm1 (Radeon RX 7900 XTX) - 24524 MiB free

sARY77 committed:
Remove obsolete HIP workaround

JohannesGaessler commented:
Is there a downside to having this call though? My understanding is that rocBLAS would be initialized anyways once it's being used so you wouldn't be saving any memory.

sARY77 committed:
Remove more references to rocBLAS

sARY77 commented:
> Is there a downside to having this call though? My understanding is that rocBLAS would be initialized anyways once it's being used so you wouldn't be saving any memory.

I was able to remove all references to rocBLAS from the code and make files. And llama-cli can still use both my GPUs. Does this mean it's no longer needed?

sARY77 committed:
Merge remote-tracking branch 'upstream/master' into Remove_obsolete_HIP_workaround

JohannesGaessler commented:
To my knowledge rocBLAS is still used internally by HIP even if it is not referenced directly. More generally, while the bug seems to have been fixed for v6.0 I have not seen confirmation that it was fixed for v5.7. And I don't think this PR would provide any benefits other than cosmetic ones. So my stance is that the workaround should be kept.

mjtalkiewicz commented:
This fixed a segfault for me when using both a 7900xtx and a 7600xt.

IMbackK commented:
The difference between calling rocblas_initialize and not calling it is that  in the rocblas_initialize case rocblas will load all tensile code objects for all operators gpus in the system while not calling rocblas_initialize will load only the object required for the current operation when that operation is first used. This means not calling rocblas_initialize avoids a small runtime cost.

As to why this should have an impact on vram usage, tensile may allocate some temporary buffers but they should not last, so i have no idea.

@mjtalkiewicz segfault is concerning, his case is special as he has a gpu for which tensile has no logic files to generate asm kernels for. In this case there is a hip fallback library that will be loaded, but this a recent addition and is afiak not tested by amd as part of the ci. in the rocblas_initialize case these will be loaded while in the non rocblas_initialize case if he only uses the xtx this will not happen

@mjtalkiewicz please describe in detail what versions of everything you are running and what gpus you are using in the test that fails. I would also try git rocblas and git tensile as the fallback stuff is pretty new.

JohannesGaessler commented:
>As to why this should have an impact on vram usage, tensile may allocate some temporary buffers but they should not last, so i have no idea.

I think it's just an issue of when these prints occur. On master they occur after initialization, with this PR they would occur after initialization.

JohannesGaessler commented:
@mjtalkiewicz How did you mean "This fixed a segfault for me when using both a 7900xtx and a 7600xt" to be interpreted? Do you mean that this PR fixed a segfault or that the workaround that this PR would removed fixed a segfault?

mjtalkiewicz commented:
@JohannesGaessler I meant this PR had fixed the segfault.

@IMbackK Updating rocblas fixed the issue, and llama.cpp is now working without this PR.

For reference, the cards were an MSI 7900 XTX and a PowerColor 7600 XT, and misbehaving version of rocblas was Fedora 41's rocblas-0:6.2.1-1

sARY77 committed:
Merge remote-tracking branch 'upstream/master' into Remove_obsolete_HIP_workaround

sARY77 committed:
Temporarily add logging of free device memory at the end of main

sARY77 commented:
@JohannesGaessler @IMbackK I added logging to confirm that with the same exact command line arguments 688 MiB more VRAM is free at the end of generation if the rocblas_initialize() call is removed. Can you please look into this again?

Wasting a significant amount of VRAM just to maintain an old workaround does not make sense to me. Even if you believe this is workaround is still needed for older cards, it should be under a command line switch that's off by default.

With rocblas_initialize():
llama_model_load_from_file_impl: using device ROCm0 (Radeon RX 7900 XTX) - 24026 MiB free
llama_model_load_from_file_impl: using device ROCm1 (Radeon RX 7900 XTX) - 24524 MiB free
...
main: using device ROCm0 (Radeon RX 7900 XTX) - 3904 MiB free
main: using device ROCm1 (Radeon RX 7900 XTX) - 3992 MiB free

Without rocblas_initialize():
llama_model_load_from_file_impl: using device ROCm0 (Radeon RX 7900 XTX) - 24524 MiB free
llama_model_load_from_file_impl: using device ROCm1 (Radeon RX 7900 XTX) - 24524 MiB free
...
main: using device ROCm0 (Radeon RX 7900 XTX) - 4248 MiB free
main: using device ROCm1 (Radeon RX 7900 XTX) - 4336 MiB free

Command line to reproduce the results:
./build/bin/llama-cli -m ./models/Llama-3.3-70B-Instruct-Q4_K_M-128K.gguf -dev ROCm0,ROCm1 -ngl 99 -c 32 -s 0 --sampling-seq k --top-k 1 -p "2 * 2 = " -n 1 -no-cnv

IMbackK commented:
Thats a bit weird and maybe a bug, i cant seam reproduce it on mi100. Regardless of the vram, issue not calling rocblas_initialize would in be a good thing to avoid the startup time cost of unused tensile objects.

However, since there are still users of rocm <6, like debian, which still packages 5.7 we should no simple remove the workaround but instead. instead call rocblas_get_version_string and match against any version with the bug (<4.0.0) and only call rocblas_initialize there.

sARY77 committed:
Merge remote-tracking branch 'upstream/master' into Remove_obsolete_HIP_workaround

sARY77 committed:
Address PR feedback

sARY77 commented:
@IMbackK I updated the PR with your recommendation. Please take a look. Thank you!

ngxson commented:
Not sure why I was marked as reviewer here, I don't know about AMD part

IMbackK commented on ggml/src/ggml-cuda/ggml-cuda.cu:
```diff
@@ -123,8 +123,15 @@ static ggml_cuda_device_info ggml_cuda_init() {
-    rocblas_initialize();
-    CUDA_CHECK(cudaDeviceSynchronize());
+    {
+        char version_string[64];
```
While this is fine it would be slightly better here to use rocblas_get_version_string_size to let rocblas tell you how big the buffer needs to be.

IMbackK commented on ggml/src/ggml-cuda/ggml-cuda.cu:
```diff
@@ -123,8 +123,15 @@ static ggml_cuda_device_info ggml_cuda_init() {
+        char version_string[64];
+        version_string[0] = '\0';
+        const rocblas_status status = rocblas_get_version_string(version_string, sizeof(version_string));
+        if (status != rocblas_status_success || version_string[0] < '4') {
```
i don't like this too much as this will ofc fail if rocblas changes its version to 10.0.0 or whatever. I think we should make a bit more effort to parse this properly. Looking at rocblas (https://github.com/ROCm/rocBLAS/blob/59825a7367a24eed4e7e8a483820592089eaf17e/library/src/buildinfo.cpp#L29) it seams we would be on the safe side to use string_split<int> here (https://github.com/ggerganov/llama.cpp/blob/df984e014714cba4c99ef894b20b51cbcef31b16/common/common.h#L439)

however currently common.h is not used outside of the clients/examples and contains code that makes no sense in the backend.
@ggerganov maybe you can weigh in on if its ok to use this header here or if we should move function somewhere else.

IMbackK reviewed (commented):
I also think we should at least issue a debug print when the workaround is in play

IMbackK commented:
> As to why this should have an impact on vram usage, tensile may allocate some temporary buffers but they should not last, so i have no idea.

Actually the extra vram use of rocblas_initialize seams to be normal and expected.
So rocblas moves the kernel code and some workspaces into vram when the tensile object initializes.
This vram cost is alos mesured by rocblas itself in its clients here https://github.com/ROCm/rocBLAS/blob/59825a7367a24eed4e7e8a483820592089eaf17e/clients/common/client_utility.cpp#L500 and indeed its about 500MB for RDNA2, while being smaller for mi100.

When we don't call rocblas_initialize rocblas will only load the code objects and workspaces for the operations actually used, so we expect this to save some vram.

So i can confirm @sARY77 observations.

sARY77 committed:
Address code review feedback

sARY77 commented:
@IMbackK I addressed your feedback to call rocblas_get_version_string_size, parse the full major version value (std::from_chars stops at the first invalid character and does not throw any exceptions), and add GGML_LOG_DEBUG in the new iteration, please take a look. Thank you!

sARY77 committed:
Merge remote-tracking branch 'upstream/master' into Remove_obsolete_HIP_workaround

IMbackK reviewed (approved):
Looks good to me now

YellowRoseCx commented:
started getting build errors on windows from this commit

sARY77 commented:
> started getting build errors on windows from this commit

@YellowRoseCx Can you please be more specific? Paste the error you are getting.

YellowRoseCx commented:
> > started getting build errors on windows from this commit
> 
> @YellowRoseCx Can you please be more specific? Paste the error you are getting.

I'm using it via a fork called KoboldCpp and my Windows Build script gives this error based off of "rocblas_get_version_string" when building via CMake:
```
AILED: CMakeFiles/ggml-rocm.dir/ggml/src/ggml-cuda/ggml-cuda.cu.obj 
C:\PROGRA~1\AMD\ROCm\6.1\bin\CLANG_~1.EXE -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DGGML_USE_CPU -DGGML_USE_CPU_AARCH64 -DGGML_USE_CUDA -DGGML_USE_HIP -DGGML_USE_HIPBLAS -DK_QUANTS_PER_ITERATION="" -DLOG_DISABLE_LOGS -DSD_USE_CUBLAS -DUSE_PROF_API=1 -D_CRT_SECURE_NO_WARNINGS -D__HIP_PLATFORM_AMD__=1 -ID:/a/koboldcpp-rocm/koboldcpp-rocm/./ggml/include -ID:/a/koboldcpp-rocm/koboldcpp-rocm/./ggml/src -ID:/a/koboldcpp-rocm/koboldcpp-rocm/./include -ID:/a/koboldcpp-rocm/koboldcpp-rocm/./otherarch -ID:/a/koboldcpp-rocm/koboldcpp-rocm/./otherarch/tools -isystem "C:/Program Files/AMD/ROCm/6.1/include" -O2 -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -std=gnu++14 -xhip -O2 -mllvm=-amdgpu-early-inline-all=true -mllvm=-amdgpu-function-calls=false -x hip --offload-arch=gfx803 --offload-arch=gfx900 --offload-arch=gfx906 --offload-arch=gfx1010 --offload-arch=gfx1011 --offload-arch=gfx1012 --offload-arch=gfx1030 --offload-arch=gfx1031 --offload-arch=gfx1032 --offload-arch=gfx1100 --offload-arch=gfx1101 --offload-arch=gfx1102 -MD -MT CMakeFiles/ggml-rocm.dir/ggml/src/ggml-cuda/ggml-cuda.cu.obj -MF CMakeFiles\ggml-rocm.dir\ggml\src\ggml-cuda\ggml-cuda.cu.obj.d -o CMakeFiles/ggml-rocm.dir/ggml/src/ggml-cuda/ggml-cuda.cu.obj -c D:/a/koboldcpp-rocm/koboldcpp-rocm/ggml/src/ggml-cuda/ggml-cuda.cu
D:/a/koboldcpp-rocm/koboldcpp-rocm/ggml/src/ggml-cuda/ggml-cuda.cu:184:17: error: no matching function for call to 'rocblas_get_version_string'
  184 |             if (rocblas_get_version_string(version.data(), version.size()) == rocblas_status_success) {
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Program Files/AMD/ROCm/6.1/include\rocblas\internal/rocblas-functions.h:19746:31: note: candidate function not viable: 1st argument ('const char *') would lose const qualifier
 19746 | ROCBLAS_EXPORT rocblas_status rocblas_get_version_string(char* buf, size_t len);
       |                               ^                          ~~~~~~~~~
D:/a/koboldcpp-rocm/koboldcpp-rocm/ggml/src/ggml-cuda/ggml-cuda.cu:187:26: error: no member named 'from_chars' in namespace 'std'
  187 |                 if (std::from_chars(version.c_str(), version.c_str() + version.length(), parsed_value).ec == std::errc()) {

      |                     ~~~~~^

2 errors generated when compiling for gfx1010.
```

JohannesGaessler commented:
I was not aware that there are compilers that will return `const char *` for `string::data` but according to [cpluplus.com](https://cplusplus.com/reference/string/string/data/) that is the spec. This can probably be fixed by using an explicit `std::vector<char>` instead.

IMbackK commented:
That part makes sense

`error: no member named 'from_chars' in namespace 'std'` however suggests the compiler is also not in C++17 mode which is probably down to not selecting it in your forked build system.

YellowRoseCx commented:
> That part makes sense
> 
> `error: no member named 'from_chars' in namespace 'std'` however suggests the compiler is also not in C++17 mode which is probably down to not selecting it in your forked build system.

It is in C++17 mode, in the cmake file is this:
```
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED true)
set(CMAKE_C_STANDARD 11)
set(CMAKE_C_STANDARD_REQUIRED true)
```

sARY77 commented:
@JohannesGaessler That site is misleading. The non-const override became standard in C++17:
https://en.cppreference.com/w/cpp/string/basic_string/data

@IMbackK I see that you already merged your change for one of the errors, but both were caused by not using C++17 mode in the compiler.

@YellowRoseCx To fix the other error please enable C++17 mode in the compiler - std::from_chars became standard in C++17:
https://en.cppreference.com/w/cpp/utility/from_chars
